mlir/lark/mlir.lark

```diff
@@ -273,10 +298,21 @@ integer_set_def     : map_or_set_id "=" integer_set_inline
 attribute_alias_def : attribute_alias "=" attribute_value
 ?definition : type_alias_def | affine_map_def | semi_affine_map_def | integer_set_def | attribute_alias_def
 
-?start : value*
-?value : module
-       | definition
-       | function
+
+// ----------------------------------------------------------------------
+// Structure of an MLIR parse-able string
+
+definition_list              : definition*
+function_list                : function*
+module_list                  : module*
+definition_and_function_list : definition_list function_list
+definition_and_module_list   : definition_list module_list
+
+mlir_file: definition_and_function_list*       -> only_functions_and_definitions_file
+         | definition_and_module_list*         -> mlir_file_as_definition_and_module_list
```

Comment:
does a function or module have to exist? can't there be files with definitions only?

Reply:
Yep, it's still optional.
```lark
definition_list              : definition*
function_list                : function*
definition_and_function_list : definition_list function_list

mlir_file: definition_and_function_list*       -> only_functions_and_definitions_file
         | definition_and_module_list*         -> mlir_file_as_definition_and_module_list
```